### PR TITLE
adds test for userInfo being applied to database encoder/decoder

### PR DESCRIPTION
**Pull request by MartinP7r**
This PR adds missing tests for #958 

I've started with a test for JSON decoding.
The test fails without the change in #958, namely:

https://github.com/groue/GRDB.swift/blob/47199ed72b6230fb498897eaaf6b4e85d1d2f2c5/GRDB/Record/FetchableRecord.swift#L123

I've used a similar structure to your other tests in that file. Please let me know if I should rather put it somewhere else.
I hope this is the type of test you were expecting or at least in the right direction.

I've submitted this as a draft, just so that you can see I'm working on it and will add the test for JSON encoding tomorrow.

### Pull Request Checklist

- [x] This pull request is submitted against the `development` branch.
- [ ] Inline documentation has been updated.
- [ ] README.md or another dedicated guide has been updated.
- [x] Changes are tested.

### CODE REVIEW

**Comment by groue** [comment]
This looks like a perfect pull request, @MartinP7r 💯

**Comment by groue** [comment]
> I've submitted this as a draft, just so that you can see I'm working on it and will add the test for JSON encoding tomorrow.

It looks like you know which test is missing better than I do. All right, I do not merge now!

File: Tests/GRDBTests/MutablePersistableRecordEncodableTests.swift [hunk 1]
```diff
@@ -800,5 +800,60 @@ extension MutablePersistableRecordEncodableTests {
...
+                } else {
+                    try container.encode(firstName, forKey: .lastName)
+                    try container.encode(lastName, forKey: .firstName)
+                }
+            }
+        }
+
+        struct StructWithNestedType : PersistableRecord, FetchableRecord, Codable {
+            static let databaseTableName = "t1"
+            static var databaseEncodingUserInfo: [CodingUserInfoKey: Any] = [CodingUserInfoKey.testKey: "correct"]
```
Comment (MartinP7r) [inline Tests/GRDBTests/MutablePersistableRecordEncodableTests.swift:830]: Writing just `static var databaseEncodingUserInfo = [CodingUserInfoKey.testKey: "correct"]` will lead to unexpected behavior. The `databaseEncodingUserInfo` dictionary will still be empty at the time it's passed to the jsonEncoder at https://github.com/groue/GRDB.swift/blob/47199ed72b6230fb498897eaaf6b4e85d1d2f2c5/GRDB/Record/FetchableRecord.swift#L123

**Comment by MartinP7r** [comment]
Thank you! 🙂
I've added the remaining test symmetrically for the userInfo logic concerning `databaseJSONEncoder`. 
With this both changes are covered as far as I can see and I hope this PR can be merged.

  
  
---

What I've learned for myself while implementing that test is how careful one has to be with Swift's `Dictionary` type-inference. 

If I write just 

```swift
static var databaseDecodingUserInfo = [CodingUserInfoKey.testKey: "correct"]
```

right here: https://github.com/groue/GRDB.swift/pull/960/commits/0c4e483e9e22df6e7b0dc5a32376f6f5e1dd76f5#r611115011

instead of explicitly naming the type like you did in your inline documentation:

```swift
static let databaseDecodingUserInfo: [CodingUserInfoKey: Any] = [decoderName: "Database"]
```

the test is going to fail and I'm guessing it's because Swift's compiler infers `[CodingUserInfoKey: String]` instead of `[CodingUserInfoKey: Any]` for the dictionary. This seems a like a small trap one should be aware of.  
I'm not even sure if this is working as intended or maybe a bug related to parameter overloading?

Sorry for the small excursion. It's not related to the PR directly.

File: Tests/GRDBTests/MutablePersistableRecordEncodableTests.swift (hunk 1 above)
Comment (groue) [reply Tests/GRDBTests/MutablePersistableRecordEncodableTests.swift:830]: Yes. Without the explicit type, the compiler sees `databaseEncodingUserInfo: [CodingUserInfoKey: String]`, which does not match the protocol requirement.

**Comment by groue** [comment]
> I've added the remaining test symmetrically for the userInfo logic concerning `databaseJSONEncoder`.
> With this both changes are covered as far as I can see and I hope this PR can be merged.

Great, thank you :-)

> This seems a like a small trap one should be aware of.

Indeed there is a note ("make sure ...") about this trap: https://github.com/groue/GRDB.swift/blob/master/README.md#the-userinfo-dictionary

**Review by groue — APPROVED** [review]

**Comment by groue** [comment]
I'll merge and ship a new version very soon :-)

**Comment by groue** [comment]
Shipped in v5.7.4

**Comment by MartinP7r** [comment]
You're not wasting any time. 🙂
Thank you! I learned a lot this weekend.

**Pull request merged**